libr/io/io_map.c

```diff
@@ -14,142 +14,81 @@
 #define CMP_END_GTE(addr, itv) \
 	(((addr) < r_itv_end (*(RInterval *)(itv))) ? -1 : 1)
 
-struct map_event_t {
-	RIOMap *map;
-	ut64 addr;
-	int id; // distinct priority in [0, len(maps))
-	bool is_to;
-};
-
-// Sort by address, (addr, is_to) precedes (addr, !is_to)
-static int _cmp_map_event(const void *a_, const void *b_) {
-	struct map_event_t *a = (void *)a_, *b = (void *)b_;
-	ut64 addr0 = a->addr - a->is_to, addr1 = b->addr - b->is_to;
-	if (addr0 != addr1) {
-		return addr0 < addr1? -1: 1;
-	}
-	if (a->is_to != b->is_to) {
-		return !a->is_to? -1: 1;
-	}
-	if (a->id != b->id) {
-		return a->id < b->id? -1: 1;
-	}
-	return 0;
-}
-
-static int _cmp_map_event_by_id(const void *a_, const void *b_) {
-	struct map_event_t *a = (void *)a_, *b = (void *)b_;
-	return a->id - b->id;
-}
-
-// Precondition: from == 0 && to == 0 (full address) or from < to
-static bool _map_skyline_push(RPVector *map_skyline, ut64 from, ut64 to, RIOMap *map) {
-	RIOMapSkyline *part = R_NEW (RIOMapSkyline), *part1;
-	if (!part) {
-		return false;
+#define CMP_END_GTE_PART(addr, part) \
+	(((addr) < (r_itv_end (((RIOMapSkyline *)(part))->itv)) || !r_itv_end (((RIOMapSkyline *)(part))->itv)) ? -1 : 1)
+
+#define CMP_BEGIN_GTE_PART(addr, part) \
+	(((addr) < (r_itv_begin (((RIOMapSkyline *)(part))->itv))) ? -1 : 1)
+
+static bool add_map_to_skyline(RIO *io, RIOMap *map) {
+	size_t slot;
+	RPVector *skyline = &io->map_skyline;
+
+	RIOMapSkyline *new_part = R_NEW (RIOMapSkyline);
+	new_part->map = map;
+	new_part->itv = map->itv;
+
+	// `part` is the first RIOMapSkyline with part->itv.addr > new_part->itv.addr
+	r_pvector_lower_bound (skyline, new_part->itv.addr, slot, CMP_BEGIN_GTE_PART);
+	RIOMapSkyline *part = slot < r_pvector_len (skyline) ? r_pvector_at (skyline, slot) : NULL;
+	if (slot) {
+		RIOMapSkyline *prev_part = r_pvector_at (skyline, slot - 1);
```

Comment:
Something is wrong. lower_bound should return first value >= than  x. Either comment, comparison operator or r_pvector_lower_bound is incorrect.

Reply:
`test_pvector_lower_bound` tests this. So that leaves one of the first two cases.

Reply:
I'd say COMP_BEGIN_GTE_PART is at fault. Passing a comparison operator which claims that either `x > x` or `x < x` to algorithm which expects a comparison operator will likely give you a wrong result at best and memory corruption in the worst case.

Just don't try to be smart and use the known correct pattern for comparing `(x > y) - (x < y)`. Not only is it correct in all cases but can even generate code that is equally good if not better than `x < y ? -1 : 1`. https://godbolt.org/z/Teo97M

Reply:
yeah, it was wrong. got confused because the other macros are for non-inclusive end, while begin is inclusive.